dp-cluster-setup-utility.py

```diff
@@ -882,6 +893,143 @@ def streamsmsgmgr_url(self):
+          <enabled>true</enabled>
+          <param>
+            <name>config.prefix</name>
+            <value>hadoop.auth.config</value>
+          </param>
+          <param>
+            <name>hadoop.auth.config.signature.secret</name>
+            <value>knox-signature-secret</value>
+          </param>
+          <param>
+            <name>hadoop.auth.config.type</name>
+            <value>kerberos</value>
+          </param>
+          <param>
+            <name>hadoop.auth.config.simple.anonymous.allowed</name>
+            <value>false</value>
+          </param>
+          <param>
+            <name>hadoop.auth.config.token.validity</name>
+            <value>1800</value>
+          </param>
+          <param>
+            <name>hadoop.auth.config.cookie.domain</name>
+            <value>HWQE.HORTONWORKS.COM</value>
+          </param>
+          <param>
+            <name>hadoop.auth.config.cookie.path</name>
+            <value>gateway/beacon-proxy</value>
+          </param>
+          <param>
+            <name>hadoop.auth.config.kerberos.principal</name>
+            <value>HTTP/{knox_host}@HWQE.HORTONWORKS.COM</value>
+          </param>
+          <param>
+            <name>hadoop.auth.config.kerberos.keytab</name>
+            <value>/etc/security/keytabs/spnego.service.keytab</value>
+          </param>
+          <param>
+            <name>hadoop.auth.config.kerberos.name.rules</name>
+            <value>DEFAULT</value>
+          </param>
+        </provider>
+     </gateway>
+     {ranger}
+     {atlas_api}
+     {beacon}
+  </topology>"""
+
+  def __init__(self, ambari, role_names, name='beacon-proxy'):
+    self.ambari = ambari
+    self.role_names = role_names
+    self.name = name
+
+  def deploy(self, knox):
+    self.update_knox_service_defs(knox)
+    template = BeaconProxyTopology.TEMPLATE.format(
+      knox_url = str(knox.base_url),
+      knox_host = self.ambari.cluster.knox_host(),
+      timestamp = int(time.time()),
+      ambari_protocol = self.ambari.base_url.protocol(),
+      ambari_host = self.ambari.internal_host,
+      ambari_port = self.ambari.base_url.port(),
+      ranger = self.ranger(),
+      atlas_api = self.atlas_api(),
+      beacon = self.beacon(),
+    )
+    return knox.add_topology(self.name, template)
+
+  def update_knox_service_defs(self, knox):
+    if 'DPPROFILER' in self.role_names:
+      stack = self.ambari.installed_stack()
+      if stack.name == 'HDP':
+        stack_version = self.ambari.current_stack_version()
+        knox.update_profiler_agent_service_def(stack_version)
+
+  def ranger(self):
+    return self.role('RANGER', self.ranger_url(), '0.1.0.0') if 'RANGER' in self.role_names else ''
+
+  def atlas_api(self):
+    return self.role('ATLAS-API', self.atlas_url(), '0.1.2.0') if 'ATLAS' in self.role_names else ''
+
+  def beacon(self):
+    return self.role('BEACON', self.beacon_url()) if 'BEACON' in self.role_names else ''
+
+  def role(self, name, url, version=''):
```

Comment:
It feels like this code is being repeated - can we do something about it?

Reply:
Yes there is code repetition. Wanted to discuss the approach for this. One approach i can think of is , create a new util class for getting ranger, atlas_api service definitions and use them in both BeaconProxyTopology and DpProxyTopology class.